### ISSUE TITLE
when using typescript recommended config it should allow for imports of js equivialent

### ISSUE THREAD

**Issue by lifeiscontent** [import-js/eslint-plugin-import#2111]
based off the conversation here: https://github.com/microsoft/TypeScript/issues/16577

if a `.ts` or `.tsx` file imports a `.js` file it will resolve `.ts` `.tsx` or `.js` but the `plugin:import/typescript` doesn't seem to have support for this functionality.

the `js` extension is used to reference files that WILL be compiled to, so they technically don't exist, but this is how the TS team is allowing people to write ESM compatible modules in TypeScript.

It would be nice to have this eslint plugin to work with this out of the box so I can enforce that all of my modules import `.js` files to make sure I'm not missing anything when exporting my package to the web.

**Comment by ljharb** [comment]
The "recommended" config of course doesn't support any typescript functionality - the default is (should always be) JavaScript.

https://github.com/benmosher/eslint-plugin-import#typescript may be helpful.

**Comment by lifeiscontent** [comment]
@ljharb Yeah, I'm specifically talking about the typescript additions.

**Comment by leepowelldev** [comment]
@lifeiscontent - try using: https://www.npmjs.com/package/eslint-import-resolver-typescript

**Comment by leepowelldev** [comment]
@ljharb - would there be any appetite for `import/extensions` rule to support a typescript option which would allow `.ts` and `.tsx` to be equivalent to a `.js` extension which the Typescript compiler allows us to use as mentioned by @lifeiscontent 

```
// foo.ts
export const foo = 'foo';

// bar.ts
import { foo } from './foo.js';
```

Currently this errors with `Missing file extension "ts" for "./foo.js"`. Maybe we could add an additional option? I'm happy to take a stab at it if it's something that was welcomed?

**Comment by ljharb** [comment]
I don’t think that would really make sense. It would be modifying a general rule with typescript-specific logic to paper over a typescript bug/flaw - one that once they fix, we’d be stuck supporting the workaround for forever.

**Comment by leepowelldev** [comment]
@ljharb - sorry, do you mind explaining what the Typescript bug/flaw is? Currently I'm struggling to be able to enforce `.js` extensions in a project using Typescript using this rule (which will go on to be published as a ESM package)

**Comment by ljharb** [comment]
@leepowelldev my understanding is that typescript explicitly forces you to omit extensions, so that it can conceptually point to `.ts` at dev time but `.js` at runtime. TS and native ESM are effectively not yet compatible, and can't be used together.

**Comment by leepowelldev** [comment]
@ljharb Typescript doesn't explicitly force you to omit extensions. It won't let you use `.ts` ot `.tsx` as an extension, but `.js` is perfectly valid (https://github.com/microsoft/TypeScript/issues/16577). This allows me to create valid ESM output from a Typescript source.

> TS and native ESM are effectively not yet compatible, and can't be used together.

I'm not sure what you mean by this.

**Comment by ljharb** [comment]
I mean that if your goal is a native ESM package (something i'd strongly discourage for many reasons, unrelated to this), then authoring in TypeScript is not yet an ideal way to get there, because of node's unfortunate "required extensions for ESM" choice.

What I'd suggest is to only use tsc as a typechecker, to use babel as your only transpiler, and then to use a babel plugin to transform the proper omitted-extensions source to whatever output you like, which can include "native ESM with extensions".

**Comment by patrickarlt** [comment]
I'm migrating a large project written in TypeScript to output native ESM modules for Node. This requires all the `import` statements in TypeScript to end in `.js` so they get output with the `.js` extensions and I would like to validate that behavior.

I don't think TypeScript is going to budge on their treatment of extensions, see https://github.com/microsoft/TypeScript/issues/16577#issuecomment-754941937, neither will Node JS back off of required extensions for ESM so it is perfectly valid to write this:

```ts
// foo.ts
export const foo = 'foo';

// bar.ts
import { foo } from './foo.js';
```

Currently this config produces `Missing file extension "ts" for "./foo.js"` which is illogical since TypeScript doesn't actually support `.ts` on file paths https://github.com/microsoft/TypeScript/issues/37582.

```js
module.exports = {
  root: true,
  parser: "@typescript-eslint/parser",
  parserOptions: {
    ecmaVersion: 2018,
    sourceType: "module"
  },
  plugins: ["@typescript-eslint", "import"],
  extends: ["plugin:import/recommended", "plugin:import/typescript"],
  rules: {
    "import/extensions": ["error", "ignorePackages"]
  },
  settings: {
    "import/extensions": [".js", ".jsx"],
    "import/parsers": {
      "@typescript-eslint/parser": [".ts", ".tsx"]
    },
    "import/resolver": {
      typescript: {
        project: "./"
      }
    }
  }
};
```

**Comment by patrickarlt** [comment]
Would a compromise maybe be to introduce a `validExentions` option so we could say that `.js` is a valid extension in `.ts` files?

**Comment by patrickarlt** [comment]
There is also this issue in`eslint-import-resolver-typescript` https://github.com/alexgorbatchev/eslint-import-resolver-typescript/issues/80 that mentions this exact use case.

**Comment by ljharb** [comment]
I think the proper place to do that is the TS resolver. It can, for example, resolve otherwise-missing `.js` files to their present `.ts` equivalent.

**Comment by leepowelldev** [comment]
Looking at the src for `extensions` rule I believe this plugin would still error even if anything was done in the resolver.

I have to disagree that supporting this feature should be done at the resolver level.

**Comment by ljharb** [comment]
That suggests to me that the extensions rule needs refactoring, if it’s hardcoding node resolution.

**Comment by leepowelldev** [comment]
I'm not sure about that ... I think it resolves just fine (would need to check), but then simply check that the file extensions match. Obviously it fails as expected when it tries to match `.js` with `.ts` or `.tsx`. We would need additional checks to allow this to be valid.

**Comment by patrickarlt** [comment]
I just opened https://github.com/alexgorbatchev/eslint-import-resolver-typescript/issues/82 so we can see what might be able to be done in the resolver. Although I do agree with @leepowelldev that it makes more sense to handle this behavior in the rules not the resolver.

This behavior is really specific to TypeScript though so would this be best as a separate rule just for TypeScript so it doesn't conflict?

**Comment by ljharb** [comment]
We don’t have any typescript-specific rules; this is a JavaScript plugin that works for TS *as well*.

**Comment by leepowelldev** [comment]
Sure, and I don't think this has to be typescript specific if we can expose a config option to support alternative relationships between extensions.
```
"import/extensions": [
  <severity>,
  "never" | "always" | "ignorePackages",
  {
    ignorePackages: true | false,
    pattern: {
      <extension>: "never" | "always" | "ignorePackages"
    },
    mapExtensions: {
      "js": "tsx?"
    }
  }
]
```
This gives the flexibility without directly supporting Typescript.

**Comment by JounQin** [comment]
This is actually already supported at https://github.com/alexgorbatchev/eslint-import-resolver-typescript/pull/56.

**Comment by ljharb** [comment]
and overrides that only apply to .ts files, that set the extension to be “always”, presumably would force .ts, and you want .js?

**Comment by JounQin** [comment]
> and overrides that only apply to .ts files, that set the extension to be “always”, presumably would force .ts, and you want .js?

Yep.

**Comment by patrickarlt** [comment]
Thanks to @JounQin I was able to get this working by excluding `eslint-import-resolver-typescript` which seems a but counter initiative but it is working. If anyone wants to take a look at the minimal reproduction I made I can push it up to GitHub.

```
// src/bar.ts - should pass
import { foo } from "./foo.js";

// src/baz.ts - should fail
import { foo } from "./foo";

// src/foo.ts
export const foo = "foo";
```

### With `eslint-import-resolver-typescript` - both `bar.ts` and `baz.ts` fail

![2021-09-09_12-22-27](https://user-images.githubusercontent.com/378557/132749711-c9087331-0a23-473b-8148-45f1cfa842d1.png)

```json
{
  "root": true,
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 12,
    "sourceType": "module"
  },
  "plugins": [
    "import"
  ],
  "rules": {
    "import/extensions": ["error", "ignorePackages"]
  },
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [
        ".ts",
        ".tsx"
      ]
    },
    "import/resolver": {
      "typescript": {
        "project": "./tsconfig.json"
      }
    }
  }
}
```

### Without `eslint-import-resolver-typescript` - works as intended

![2021-09-09_12-22-14](https://user-images.githubusercontent.com/378557/132749724-6dd7916f-8eff-4513-bdcd-e3a771db3804.png)

```json
{
  "root": true,
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 12,
    "sourceType": "module"
  },
  "plugins": [
    "import"
  ],
  "rules": {
    "import/extensions": ["error", "ignorePackages"]
  },
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [
        ".ts",
        ".tsx"
      ]
    }
  }
}
```

**Comment by leepowelldev** [comment]
@patrickarlt so if we just **don't** use `eslint-import-resolver-typescript` then this works?

**Comment by patrickarlt** [comment]
@leepowelldev It works unless you also enable the `import/no-unresolved` rule. In my quest to make a VERY minimal reproduction I didn't extend `plugin:import/recommended` and `plugin:import/typescript`. So if you *don't* use `eslint-import-resolver-typescript` and keep `import/no-unresolved` disabled it works which I don't really think is intended. I think this is what was discussed in https://github.com/import-js/eslint-plugin-import/issues/2111#issuecomment-916212917. The resolver works but the `import/extensions` rule doesn't think the extension is correct.

I think at this point my minimal reproduction would be helpful https://github.com/patrickarlt/minimal-typescript-eslint-with-extensions.

### Adding `import/no-unresolved` without  `eslint-import-resolver-typescript` breaks resolving (predictable)

![2021-09-09_12-50-07](https://user-images.githubusercontent.com/378557/132753113-aee2d298-1c48-49dd-933b-f01b66d3e2b1.png)

```json
{
  "root": true,
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 12,
    "sourceType": "module"
  },
  "plugins": [
    "import"
  ],
  "extends": [
    "plugin:import/recommended",
    "plugin:import/typescript"
  ],
  "rules": {
    "import/extensions": ["error", "ignorePackages"]
  },
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [
        ".ts",
        ".tsx"
      ]
    }
  }
}
```

### Adding `import/no-unresolved` + `eslint-import-resolver-typescript` breaks (incorrect?)

![2021-09-09_12-53-07](https://user-images.githubusercontent.com/378557/132753542-67e57570-1147-4280-b622-0282bada855e.png)

```json
{
  "root": true,
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 12,
    "sourceType": "module"
  },
  "plugins": [
    "import"
  ],
  "extends": [
    "plugin:import/recommended",
    "plugin:import/typescript"
  ],
  "rules": {
    "import/extensions": ["error", "ignorePackages"]
  },
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [
        ".ts",
        ".tsx"
      ]
    },
    "import/resolver": {
      "typescript": {
      }
    }
  }
}
```

**Comment by ljharb** [comment]
Unfortunately `no-unresolved` is necessary to detect missing *untyped* packages, but perhaps adding an option to `no-unresolved` might be the solution here?

**Comment by patrickarlt** [comment]
@ljharb The behavior I think everyone is asking for is this:

Using this config:

```json
{
  "root": true,
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 12,
    "sourceType": "module"
  },
  "plugins": [
    "import"
  ],
  "extends": [
    "plugin:import/recommended",
    "plugin:import/typescript"
  ],
  "rules": {
    "import/extensions": ["error", "ignorePackages"]
  },
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [
        ".ts",
        ".tsx"
      ]
    },
    "import/resolver": {
      "typescript": {
      }
    }
  }
}
```

And these files

```ts
// src/bar.ts - should pass
import { foo } from "./foo.js";

// src/baz.ts - should fail
import { foo } from "./foo";

// src/foo.ts
export const foo = "foo";
```

The expected output should be:

![](https://user-images.githubusercontent.com/378557/132749724-6dd7916f-8eff-4513-bdcd-e3a771db3804.png)

Currently it is: 

![](https://user-images.githubusercontent.com/378557/132753542-67e57570-1147-4280-b622-0282bada855e.png)

This is because `import { foo } from "./foo.js";` is valid TypeScript that compiles to `import { foo } from "./foo.js";` If you want the extensions to output from the TypeScript compiler you must add them in the source.

**Comment by JounQin** [comment]
@patrickarlt Try to add `ts: never, tsx: never`.

**Comment by patrickarlt** [comment]
@JounQin unfortunately that just seem to make everything pass. 

![2021-09-09_20-17-56](https://user-images.githubusercontent.com/378557/132793984-d4dd1b89-7e4b-447b-9eff-6a1656458b4d.png)

```json
{
  "root": true,
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 12,
    "sourceType": "module"
  },
  "plugins": [
    "import"
  ],
  "extends": [
    "plugin:import/recommended",
    "plugin:import/typescript"
  ],
  "rules": {
    "import/extensions": [
      "error",
      "always",
      {
        "ts": "never",
        "tsx": "never"
      }
    ] 
  },
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [
        ".ts",
        ".tsx"
      ]
    },
    "import/resolver": {
      "typescript": {
      }
    }
  }
}
```

**Comment by JounQin** [comment]
@patrickarlt 

I've already said previously.

> @ljharb The problem is there is no way to enforce to use `.js` extension currently for `.ts` files.

**Comment by beamery-tomht** [comment]
Any movement on this? It would be great if we can get the extension linting working correctly for ESM in TS as well as the `no-unresolved`. I'm experiencing exactly the same issues as described by @patrickarlt

**Comment by ljharb** [comment]
TS recently merged a PR which might address this problem in TS itself.

**Comment by beamery-tomht** [comment]
@ljharb can you elaborate on that please? which PR and how does it solve the problem? The request here is for a linting rule, we already have the ability to import TS files using a `.js` extension for ESM build outputs. The linting rule would ensure that the extension isn't forgotten

**Comment by ljharb** [comment]
I mean https://github.com/microsoft/TypeScript/pull/45884 - which I believe will allow TS to do the right thing here, so that your imports in code don't have to have extensions (since source code shouldn't be using them) and they'll still transpile/resolve to the proper extensioned path.

In other words, a solution here would be a hack around a TypeScript flaw; TS seems on track to fix that flaw, obviating the need for any workaround.

**Comment by beamery-tomht** [comment]
From my understanding, the extension is mandatory in both Node and the Browser https://nodejs.org/api/esm.html#esm_mandatory_file_extensions

It's only bundling systems like babel/webpack that have added the feature of omitting the extension.

TS is not automating this extension as it's an assumption of the extension. TS won't modify import statements at compile.

Consider the following:

```ts
// file.ts
import x from './x';
```

Given

```
somedir/
  file.ts
  x.mjs
  x.d.ts
  x.ts
  x.js
  x.jsx
  x.tsx
```

So I wouldn't call it a "hack". I don't see them changing this behaviour. 

It would make everyone happy to be able to lint against this extension being present in relative imports.

**Comment by ljharb** [comment]
Your understanding is incorrect; the browser has no concept of extensions whatsoever, and the extension in node is only mandatory for relative imports - imports from packages should omit them by setting up extensionless entry points in their “exports” field.

**Comment by beamery-tomht** [comment]
Perhaps I misunderstood this statement about mandatory extensions in native modules. I took "native" here to mean native in the browser.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules#:~:text=Note%3A%20In%20some%20module%20systems%2C%20you%20can%20omit%20the%20file%20extension%20and%20the%20leading%20/%2C%20./%2C%20or%20../%20(e.g.%20%27modules/square%27).%20This%20doesn%27t%20work%20in%20native%20JavaScript%20modules.

It states clearly that the server would need to support the mime type, but the browser will infer the content type from the extension.

**Comment by ljharb** [comment]
That is explicitly saying that browsers provide no "lookup" - browsers in NO WAY have the concept of extensions, and the MIME type is only and always explicitly defined by the server and/or by the HTML tag that loaded it.

**Comment by leepowelldev** [comment]
Not quite sure where this conversation is going. Personally I don't believe the upcoming TS release will address this issue. @ljharb obviously advocates using extensionless imports, while a number of others want to use them (for whatever personal reasons they may have).

**Comment by tpluscode** [comment]
> @ljharb obviously advocates using extensionless imports

Wait,  what? Extensionless imports only work for bare module and explicit declarations in package.json `exports` field. Browser argument is rather a little beside the point. Yes, browser does not care for extension per-se but if the module is `/foo/bar.js` then it cannot be imported as `/foo/bar`. 

So unless TS has the option to output explicit extensions in imported paths we are forced to make sure all our TS modules are imported as `*.js`

**Comment by leepowelldev** [comment]
@tpluscode - I completely agree, but I think @ljharb made the argument that the `.js` extension should be added as part of the build process, either by a babel plugin or as subsequent step to the typescript compiler.

**Comment by tpluscode** [comment]
That's not so simple. An `import from `./foo/bar` can mean both `./foo/bar.js` as well as `./foo/bar/index.js` and both can exist on disk. Unless the import module is explicit, trying to defer "fixing" paths to the build is bound to be error prone. This is exactly where a import linter seems to fit right in

**Comment by ljharb** [comment]
This plugin should certainly disambiguate those possibilities, but that actually happens very rarely in practice.

**Comment by ljharb** [comment]
@beamery-tomht they already do that with `.ts` to `.js` on the file; they just don't want to modify your source. So, if your source remains extensionless, then https://github.com/microsoft/TypeScript/pull/45774 (merged in https://github.com/microsoft/TypeScript/pull/45884) i believe should work, unless i'm misunderstanding something.

**Comment by beamery-tomht** [comment]
They aren't modifying the extension at compile time. They are supporting the use of extensions in your source code. When the source is extensionless, the dist is also extensionless, which is invalid ESM.

**Comment by ljharb** [comment]
No, it is not. It's saying that the browser will ask the webserver for the exact URL in the source code, whether it has an extension or not, and that it's up to the webserver to handle it. It's exceedingly easy - and in fact common - for webservers to do extension lookup, and certainly to do `index` lookup.

Browsers do not have any idea of what an extension *is* - the word is meaningless in a browser context.

**Comment by leepowelldev** [comment]
> they already do that with `.ts` to `.js` on the file; they just don't want to modify your source. So, if your source remains extensionless, then [microsoft/TypeScript#45774](https://github.com/microsoft/TypeScript/pull/45774) (merged in [microsoft/TypeScript#45884](https://github.com/microsoft/TypeScript/pull/45884)) i believe should work, unless i'm misunderstanding something.

This is part of the problem - TS doesn’t allow the use of .ts and .tsx extensions in source code. So you either have to be extensionless or use .js

**Comment by leepowelldev** [comment]
I’ll be honest, I don’t even remember what was being asked for anymore 😂

If it’s that we can use (and enforce) .js extensions via the import/extensions rule without unresolved errors that’d be great.

**Comment by ljharb** [comment]
Right - it seems like you can already get that by using `.js` in your source, and running your build process, at which point the linter will be able to resolve those files.

**Comment by JounQin** [comment]
They are already been resolved correctly, the problem here is we want to enforce to use `.js` extension in `.ts` source codes.

**Comment by leepowelldev** [comment]
Just looking at the TS 4.5 release notes, and it looks like (unless I'm interpreting it incorrectly) that the `.js` extensions will need to be manually added to support the new `node12` and `nodenext` targets.

IMO, these two new targets make getting this issue resolved even more important.

**Comment by AndrewLeedham** [comment]
@ljharb If I were to add an option say `extensionOverride` in a PR to support this functionality, is that a feature that would land in `eslint-plugin-import`. Happy to give it a shot just don't want to start the work if you don't see it making sense in this plugin?

Thinking it would work like this:
```javascript
'import/extensions': [
  'error',
  {
    ts: { mode: 'always', extensionOverride: 'js' },
  },
]
```

**Comment by Haroenv** [comment]
What I did in my project (https://github.com/algolia/instantsearch.js) was:

1. no extensions inside the project, enforced by this plugin
2. add extensions using a custom babel plugin: https://github.com/algolia/react-instantsearch/blob/master/scripts/babel/extension-resolver.js
3. test using node (which is the most strict) that all extensions were added: https://github.com/algolia/react-instantsearch/blob/master/test/module/packages-are-es-modules.mjs

This way all tooling works as before, and I get errors if an extension is mistakenly added or if it didn't resolve correctly.

**Comment by AndrewLeedham** [comment]
@Haroenv thanks for sharing that, looks like a great approach!

@ljharb Fair enough I will leave it for now then.

**Comment by manovotny** [comment]
I just arrived on scene here yesterday after adding this plugin to our first project using TypeScript and trying to output ESM. It is disheartening to see postering and personal opinion when people are just trying to make it easy to do the right thing in their codebases, at least according to the TypeScript and ESM specs. 

Is there any way to refocus and move past opinions and personal preferences toward a meaningful and actionable path forward, again, at least according to the TypeScript and ESM specs?

**Comment by ljharb** [comment]
@manovotny "the right thing" is always a personal opinion/preference. There's no TS spec, and the ESM spec does not in any way mention specifiers, so the only things that matter are the behavior of browsers (exact URLs only, extensions doesn't exist as a concept), node (extensions are required for relative paths; "exports" means they don't have to be required for anything imported from a package), and babel/typescript (ESM uses the CJS algorithm; this is what almost all code authored with ESM is currently using).

**Comment by eturino** [comment]
I am probably adding noise (apologies in advance), but I may have missed the solution to my problem.
I have a TS for node codebase, which needs to use ESM now. As such, I was following this:

https://www.typescriptlang.org/docs/handbook/esm-node.html
which among other things say: 
> relative import paths need full extensions (we have to write `import "./foo.js"` instead of `import "./foo"`)

Currently if I do that I get a linter error with
> ESLint: Missing file extension "ts" for "./foo.js"(import/extensions)

now, foo will be a `foo.ts` file indeed. And in fact, if I try to say `import "./foo.ts"` I get a TS error 
> TS2691: An import path cannot end with a '.ts' extension. Consider importing './foo.js' instead.

I could turn off import/extensions for all ts files to avoid the warning, but the thing is that I need a linter rule to let me know which extensions are missing, since the ones that are missing will end up not working at run time.
> Error [ERR_MODULE_NOT_FOUND]: Cannot find module 'Somewhere/foo' imported from /Somewhere/index.js

The solution to allow an optional override of extensions config to say to the linter "it's a .ts extension, we're cool if you import .js instead" seems like a good solution, but I saw in your message https://github.com/import-js/eslint-plugin-import/issues/2111#issuecomment-1023290190 @ljharb that it doesn't seem right to you.

Am I missing a solution that will get me warnings of missing extensions but still compile right?

**Comment by ljharb** [comment]
@tilgovi sure. but if you want your source to include extensions, it should include `.ts`, and your build should transform that to `.mjs`.

**Comment by andersk** [comment]
It’s reasonable that you want to avoid adding TypeScript-specific logic to this JavaScript-centered plugin. What do you think about making the issue become `eslint-import-resolver-typescript`’s problem, as follows?

Extend the resolver plugin API to let it return two paths, an import path and a physical path. Rules like `extensions` that examine the filename itself would work against the import path, while rules like `no-unresolved` that examine the file’s contents would read it through the physical path.

By default the import and physical paths would be the same, but `eslint-import-resolver-typescript` could be modified to return `foo.js` for the import path and `foo.ts` for the physical path, to match `tsc`’s expectations. (This would be controlled by an `eslint-import-resolver-typescript` option of course, so that it could still be used in Babel-like configurations.)

Is this a solution that would make everyone happy?

**Comment by ljharb** [comment]
That definitely seems like an appropriate extension to the API, and would probably solve a lot of issues in TS where the physical path is a d.ts file but the import path is not.

How do you propose extending that in a non-breaking way?

**Comment by tilgovi** [comment]
Is an extension to the API needed? The current resolve interface returns the physical path, and it's called with the import path. Both paths are known, then.

**Comment by JounQin** [comment]
@tilgovi Maybe you can draft a PR for it.

**Comment by tilgovi** [comment]
What need is not being met by the TypeScript resolver today?

**Comment by JounQin** [comment]
> What need is not being met by the TypeScript resolver today?

TS resolver is working as expected, this plugin does handle extensions correctly.

**Comment by JounQin** [comment]
> import path rather than the resolved path.

I totally agree with this, but it could be a BREAKING change.

**Comment by tilgovi** [comment]
The extensions rule knows both the import path and the resolved path (if it resolves). It needs to use the resolver to support checking whether the resolved path is a package or not. 

Currently, when the path resolves, the extensions rule is getting the expected extension from the resolved path. The error happens because the resolved `.ts` file doesn't end in the expected `.js`. 

The TypeScript team has made it clear that they don't think TypeScript should transform paths during compilation. They are considering making it possible to reference `.ts` files in the source, though (microsoft/TypeScript#37582). They're not going to transform `.ts` into `.js` in this mode, but expect to use it for compilation targets where the build output is not JavaScript and the runtime isn't Node.js such that the `.ts` files will actually be present at runtime. Things are even weirder when there are bundlers involved, because there's actually no files imported at runtime.

So we need to be careful not to haphazardly expand the role of the resolver. Let's take a step back and ask what the user expects the extensions error to do. Are they trying to prevent a runtime error? Are they trying to prevent a bundler error? A compilation error? Are they just enforcing a stylistic preference?

**Comment by tilgovi** [comment]
> So we need to be careful not to haphazardly expand the role of the resolver.

For example, just because the resolver might be able to tell us what file should resolve to at runtime doesn't mean that's what the user wants their source to say. That may be what TypeScript users want today for a Node.js runtime target, but it might not be what someone else wants today or tomorrow.

**Comment by tilgovi** [comment]
If we do want to assume that it's the resolver responsibility to error if a source path is incorrect, we could make the change I was suggesting and use the import path's extension should or should not be present given the options of the extensions rule, as I said above.

As @JounQin said, that'd be a breaking change. And as I said, that'd be enshrining a particular role for the resolver that may or may not be quite the same as its original purpose. It would be saying that the resolver's role is not just to help this plugin find files to check, but to say whether or not the import is correct. That's a bit different than a stylistic check, though. That's saying it's the resolvers responsibility to hard error when the user writes `.ts` and should have written `.js`. Instead of resolving the `.ts` file, it should say no, `./foo.ts` does not resolve.

**Comment by JounQin** [comment]
@tilgovi 

The new behavior can be enabled under a new setting flag, then there will be no breaking change.

If you can contribute to draft a PR, that would be great!

**Comment by tilgovi** [comment]
I'll sleep on it first. I'm not sure how to name the flag or communicate what it does yet. Any suggestions are welcome.

**Comment by JounQin** [comment]
> I'll sleep on it first. I'm not sure how to name the flag or communicate what it does yet. Any suggestions are welcome.

Good night. How about `preferImportExtension`?

**Comment by justingrant** [comment]
@JounQin @tilgovi Was there a conclusion from the discussion above about a "desired file extension" for TS relative imports?

BTW, another thing that could be helpful about this option would be if this plugin were extended in the future to support auto-fix of this problem by adding missing extensions... although for non-TS files I'm not sure how valuable that would be.

**Comment by jordansexton** [comment]
FWIW, I wrote an eslint plugin that requires js extensions, and also fixes them: https://github.com/solana-labs/eslint-plugin-require-extensions

**Comment by tpluscode** [comment]
> FWIW, I wrote an eslint plugin that requires js extensions, and also fixes them: https://github.com/solana-labs/eslint-plugin-require-extensions

Finally, the savior has arrived. Thank you for that

**Comment by leppaott** [comment]
> FWIW, I wrote an eslint plugin that requires js extensions, and also fixes them: https://github.com/solana-labs/eslint-plugin-require-extensions

This works but this plugin's `ignorePackages` would be needed to be usable. A shame this hasn't been resolved yet.

**Comment by soryy708** [comment]
Could not reproduce on up to date versions.

Note: [TypeScript allowImportingTsExtensions](https://www.typescriptlang.org/tsconfig/#allowImportingTsExtensions)

**Comment by JounQin** [comment]
> Could not reproduce on up to date versions.
> 
> Note: [TypeScript allowImportingTsExtensions](https://www.typescriptlang.org/tsconfig/#allowImportingTsExtensions)

No everyone uses that option.
